lms/djangoapps/courseware/access_response.py

```diff
@@ -0,0 +1,116 @@
+        return self.has_access
+
+    def to_json(self):
+        """
+        Creates a serializable JSON representation of an AccessResponse object.
+
+        Returns:
+            dict: JSON representation
+        """
+        return {
+            "has_access": self.has_access,
+            "error_code": self.error_code,
+            "developer_message": self.developer_message,
+            "user_message": self.user_message
+        }
+
+
+class AccessError(AccessResponse):
+    """
+    Class that holds information about the error in the case of an access denial in has_access.
+    Contains the error code, user and developer messages. Subclasses represent specific errors.
+    """
+    def __init__(self, error_code, developer_message, user_message):
+        """
+        Creates an AccessError object.
+
+        An AccessError object is represents an AccessResponse where access is denied (has_access is False).
+
+        Arguments:
+            error_code (String): unique identifier for the specific type of error
+            developer_message (String): message to show the developer
+            user_message (String): message to show the user
+
+        """
+        super(AccessError, self).__init__(False, error_code, developer_message, user_message)
+
+
+class StartDateError(AccessError):
+    """
+    Access denied because the course has not started yet and the user is not staff
+    """
+    def __init__(self, start_message):
+        error_code = "course_not_started"
+        developer_message = (
+            "Course does not start until {start}"
+        ).format(start=start_message)
+        user_message = start_message
+        super(StartDateError, self).__init__(error_code, developer_message, user_message)
+
+
+class MilestoneError(AccessError):
+    """
+    Access denied because the user hasn't completed the needed milestones (pre-reqs/entrance exams).
+    """
+    def __init__(self):
+        error_code = "unfulfilled_milestones"
+        developer_message = "User has not completed the necessary milestones"
+        user_message = "You have uncompleted milestones"
+        super(MilestoneError, self).__init__(error_code, developer_message, user_message)
+
+
+class VisibilityError(AccessError):
+    """
+    Access denied because the user does have the correct role to view this course.
+    """
+    def __init__(self):
+        error_code = "not_visible_to_user"
+        developer_message = "Course is not visible to this user"
+        user_message = "Course is not visible to you"
+        super(VisibilityError, self).__init__(error_code, developer_message, user_message)
+
+
+class MobileAvailabilityError(AccessError):
+    """
+    Access denied because the course is not available on mobile for the user
+    """
+    def __init__(self):
+        error_code = "mobile_unavailable"
+        developer_message = "Course is not available on mobile for this user"
+        user_message = "You cannot view this course on mobile"
```

Comment:
you're right for both, a user shouldn't see this message, but as the user_message field is already there I put one in